### PR TITLE
feat: optimize the dictionaries in the output of deduplicate node (#1827)

### PR DESCRIPTION
**Pull request by tustvold**
Wires in the dictionary optimization added in #1830 to the RecordBatchDeduplicator.

This may become partly redundant once #1828 is implemented, but not entirely as this ensures that the dictionaries don't contain unused values. At  least until the SortPreservingMerge is updated to compute accurate dictionaries, likely part of https://github.com/apache/arrow-rs/issues/506, I think it will be necessary to recompute them either in RecordBatchDeduplicator or a separate ExecutionPlan.

FYI @NGA-TRAN

### CODE REVIEW

**Review by e-dard — APPROVED** [review]
This LGTM 👍

File: query/src/provider/deduplicate.rs [hunk 1]
```diff
@@ -777,6 +779,78 @@ mod test {
...
+        let results = dedupe(vec![batch1, batch2], sort_keys).await;
+
+        let col = results.output[0]
+            .column(results.output[0].schema().column_with_name("t1").unwrap().0)
+            .as_any()
+            .downcast_ref::<DictionaryArray<Int32Type>>()
+            .unwrap();
+
+        // Should produce optimised dictionaries
+        assert_eq!(col.keys().len(), 2);
```
Comment (alamb) [inline query/src/provider/deduplicate.rs:836]: Is this line the main point of this test? Namely  testing that the `a, a` values in have a single dictionary entry?

Is it worth asserting that the second output batch has two dictionary keys (`b` and `c`)?

File: query/src/provider/deduplicate/algo.rs [hunk 2]
```diff
@@ -233,7 +234,8 @@ impl RecordBatchDeduplicator {
             .map(|old_column| old_column.slice(offset, len))
             .collect();
 
-        RecordBatch::try_new(schema, new_columns)
+        let batch = RecordBatch::try_new(schema, new_columns)?;
+        optimize_dictionaries(&batch)
```
Comment (alamb) [inline query/src/provider/deduplicate/algo.rs:238]:
```suggestion
        // At time of writing, `concat_batches` concatenates the
        // contents of dictionaries as well; Do a post pass to remove the
        // redundancy if possible
        optimize_dictionaries(&batch)
```

**Review by alamb — APPROVED** [review]
Code looks good to me. :shipit:

**Pull request merged**